extensions/sql-assessment/src/engine.ts

```diff
@@ -0,0 +1,175 @@
+
+
+	constructor(service: mssql.ISqlAssessmentService) {
+		this.sqlAssessment = service;
+	}
+
+	public get isServerConnection(): boolean {
+		return !this.connectionProfile.databaseName || this.connectionProfile.databaseName === 'master';
+	}
+	public get databaseName(): string {
+		return this.connectionProfile.databaseName;
+	}
+	public get recentResult(): SqlAssessmentResultInfo {
+		return this.lastInvokedResults;
+	}
+	public get targetName(): string {
+		return this.isServerConnection ? this.connectionProfile.serverName : this.connectionProfile.databaseName;
+	}
+
+
+	public async initialize(connectionId: string) {
+		this.connectionUri = await azdata.connection.getUriForConnection(connectionId);
+		this.connectionProfile = await azdata.connection.getCurrentConnection();
+	}
+
+	public async performAssessment(asmtType: AssessmentType, onResult: OnResultCallback): Promise<void> {
+		if (this.isServerConnection) {
+			await this.performServerAssessment(asmtType, onResult);
+		} else {
+			if (asmtType === AssessmentType.AvailableRules) {
+				await onResult(await this.sqlAssessment.getAssessmentItems(this.connectionUri, azdata.sqlAssessment.SqlAssessmentTargetType.Database), asmtType, false);
+			} else {
+				const result = await this.sqlAssessment.assessmentInvoke(this.connectionUri, azdata.sqlAssessment.SqlAssessmentTargetType.Database);
+
+				this.lastInvokedResults = {
+					connectionInfo: this.connectionProfile,
+					dateUpdated: Date.now(),
+					result: result
+				};
+
+				await onResult(result, asmtType, false);
+				if (asmtType === AssessmentType.InvokeAssessment) {
+					this.saveAssessment(this.databaseName, result);
+				}
+			}
+		}
+
+		if (asmtType === AssessmentType.InvokeAssessment && this.historicalRecords !== undefined) {
+			this.historicalRecords.push({
+				result: this.lastInvokedResults.result,
+				dateUpdated: this.lastInvokedResults.dateUpdated
+			});
+		}
+	}
+
+	public generateAssessmentScript(): Promise<azdata.ResultStatus> {
+		//TelemetryReporter.sendActionEvent(SqlAssessmentTelemetryView, SqlTelemetryActions.ExportAssessmentResults);
+		return this.sqlAssessment.generateAssessmentScript(this.lastInvokedResults.result.items, '', '', azdata.TaskExecutionMode.script);
+	}
+
+	public async readHistory(): Promise<SqlAssessmentRecord[]> {
+		if (this.historicalRecords === undefined) {
+			await this.loadHistory();
+		}
+
+		return this.historicalRecords;
+	}
+
+
+
+	private async loadHistory(): Promise<void> {
+		this.historicalRecords = [];
+		const files: TargetWithChildren[] = await readHistoryFileNames(this.targetName);
+
+		for (let nFileName = 0; nFileName < files.length; nFileName++) {
+			const file: TargetWithChildren = files[nFileName];
+			const content: string = await fs.readFile(file.target, 'utf8');
+			const result: azdata.SqlAssessmentResult = JSON.parse(content);
+
+			for (let nChild = 0; nChild < file.children.length && this.isServerConnection; nChild++) {
```

Comment:
Seems odd to check isServerConnection each iteration - is that ever going to change mid-execution?